File: apps/mobile/lib/models/recipe.dart
import '../utils/quantity_text.dart';
import 'sync_metadata.dart';

class RecipeIngredient {
  final String name;
  final String quantity;
  final String unit;
  final String amount;

  RecipeIngredient({
    required this.name,
    this.quantity = '',
    this.unit = '',
    String? amount,
  }) : amount = amount ?? _composeAmount(quantity, unit);

  static String _composeAmount(String quantity, String unit) {
    final q = quantity.trim();
    final u = unit.trim();
    if (q.isEmpty && u.isEmpty) return '';
    if (q.isEmpty) return u;
    if (u.isEmpty) return q;
    return '$q$u';
  }

  static _LegacyAmountParts _parseLegacyAmount(String amount) {
    final trimmed = amount.trim();
    if (trimmed.isEmpty) return const _LegacyAmountParts('', '');
    final parsed = parseLeadingQuantity(trimmed);
    if (parsed == null) {
      return _LegacyAmountParts('', trimmed);
    }
    return _LegacyAmountParts(parsed.magnitude, parsed.remainder);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RecipeIngredient &&
          runtimeType == other.runtimeType &&
          name == other.name &&
          quantity == other.quantity &&
          unit == other.unit &&
          amount == other.amount;

  @override
  int get hashCode => Object.hash(name, quantity, unit, amount);

  RecipeIngredient copyWith({
    String? name,
    String? quantity,
    String? unit,
    String? amount,
  }) {
    final preservedAmount =
        amount ?? (quantity == null && unit == null ? this.amount : null);
    return RecipeIngredient(
      name: name ?? this.name,
      quantity: quantity ?? this.quantity,
      unit: unit ?? this.unit,
      amount: preservedAmount,
    );
  }

  /// Whether this ingredient carries a plain numeric magnitude that [scaledBy]
  /// can multiply. False for measures like "适量"/"少许" or ranges like "1-2".
  bool get isScalable => double.tryParse(quantity.trim()) != null;

  /// Returns this ingredient with its numeric magnitude multiplied by [factor],
  /// re-composing the display amount from the scaled quantity + unit.
  ///
  /// A [factor] of 1 is a no-op that preserves any explicit amount. Ingredients
  /// whose quantity is not a plain number (e.g. "适量", or a range like "1-2")
  /// are returned unchanged — scaling a non-numeric measure would be meaningless.
  RecipeIngredient scaledBy(double factor) {
    if (factor == 1) return this;
    final magnitude = double.tryParse(quantity.trim());
    if (magnitude == null) return this;
    return RecipeIngredient(
      name: name,
      quantity: formatQuantity(magnitude * factor),
      unit: unit,
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'quantity': quantity, 'unit': unit, 'amount': amount};
  }

  factory RecipeIngredient.fromJson(Map<String, dynamic> json) {
    final amount = json['amount'] as String? ?? '';
    final hasNewShape =
        json.containsKey('quantity') || json.containsKey('unit');
    if (hasNewShape) {
      return RecipeIngredient(
        name: json['name'] as String? ?? '',
        quantity: json['quantity'] as String? ?? '',
        unit: json['unit'] as String? ?? '',
        amount: amount,
      );
    }
    final parts = _parseLegacyAmount(amount);
    return RecipeIngredient(
      name: json['name'] as String? ?? '',
      quantity: parts.quantity,
      unit: parts.unit,
      amount: amount,
    );
  }
}

class _LegacyAmountParts {
  const _LegacyAmountParts(this.quantity, this.unit);
  final String quantity;
  final String unit;
}

class Recipe {
  final String id;
  final String name;
  final String category;
  final int difficulty;
  final int cookingMinutes;
  final String description;
  final List<RecipeIngredient> ingredients;
  final List<String> steps;
  final List<String> tags;
  final String? imageUrl;
  final int remoteVersion;
  final DateTime? clientUpdatedAt;
  final DateTime? deletedAt;

  SyncMetadata get syncMetadata => SyncMetadata(
    remoteVersion: remoteVersion,
    clientUpdatedAt: clientUpdatedAt,
    deletedAt: deletedAt,
  );

  const Recipe({
    required this.id,
    required this.name,
    required this.category,
    required this.difficulty,
    required this.cookingMinutes,
    required this.description,
    required this.ingredients,
    required this.steps,
    this.tags = const [],
    this.imageUrl,
    this.remoteVersion = 0,
    this.clientUpdatedAt,
    this.deletedAt,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Recipe && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  Recipe copyWith({
    String? id,
    String? name,
    String? category,
    int? difficulty,
    int? cookingMinutes,
    String? description,
    List<RecipeIngredient>? ingredients,
    List<String>? steps,
    List<String>? tags,
    String? imageUrl,
    int? remoteVersion,
    DateTime? clientUpdatedAt,
    DateTime? deletedAt,
    bool clearClientUpdatedAt = false,
    bool clearDeletedAt = false,
  }) {
    return Recipe(
      id: id ?? this.id,
      name: name ?? this.name,
      category: category ?? this.category,
      difficulty: difficulty ?? this.difficulty,
      cookingMinutes: cookingMinutes ?? this.cookingMinutes,
      description: description ?? this.description,
      ingredients: ingredients ?? this.ingredients,
      steps: steps ?? this.steps,
      tags: tags ?? this.tags,
      imageUrl: imageUrl ?? this.imageUrl,
      remoteVersion: remoteVersion ?? this.remoteVersion,
      clientUpdatedAt: clearClientUpdatedAt
          ? null
          : clientUpdatedAt ?? this.clientUpdatedAt,
      deletedAt: clearDeletedAt ? null : deletedAt ?? this.deletedAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'category': category,
      'difficulty': difficulty,
      'cookingMinutes': cookingMinutes,
      'description': description,
      'ingredients': ingredients.map((e) => e.toJson()).toList(),
      'steps': List<String>.from(steps),
      'tags': List<String>.from(tags),
      'imageUrl': imageUrl,
      'remoteVersion': remoteVersion,
      'clientUpdatedAt': dateTimeToJsonValue(clientUpdatedAt),
      'deletedAt': dateTimeToJsonValue(deletedAt),
    };
  }

  factory Recipe.fromJson(Map<String, dynamic> json) {
    return Recipe(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      category: json['category'] as String? ?? '',
      difficulty: (json['difficulty'] as num?)?.toInt() ?? 0,
      cookingMinutes: (json['cookingMinutes'] as num?)?.toInt() ?? 30,
      description: json['description'] as String? ?? '',
      ingredients:
          (json['ingredients'] as List<dynamic>?)
              ?.whereType<Map<String, dynamic>>()
              .map((e) => RecipeIngredient.fromJson(e))
              .toList() ??
          const [],
      steps:
          (json['steps'] as List<dynamic>?)?.whereType<String>().toList() ??
          const [],
      tags:
          (json['tags'] as List<dynamic>?)?.whereType<String>().toList() ??
          const [],
      imageUrl: json['imageUrl'] as String?,
      remoteVersion: (json['remoteVersion'] as num?)?.toInt() ?? 0,
      clientUpdatedAt: dateTimeFromJsonValue(json['clientUpdatedAt']),
      deletedAt: dateTimeFromJsonValue(json['deletedAt']),
    );
  }
}

extension RecipeDifficultyLabel on Recipe {
  String get difficultyLabel {
    if (difficulty <= 0) {
      return '难度未设置';
    }

    final level = difficulty.clamp(1, 5).toInt();
    return '难度 $level/5';
  }
}

File: apps/ios/FreshPantry/Domain/Models/Recipe.swift
import Foundation

/// Sub-value inside `Recipe.ingredients`. Value-equal over name/quantity/unit/
/// amount. `amount` is derived from quantity+unit when not supplied (mirrors the
/// Dart constructor's `_composeAmount`).
struct RecipeIngredient: Equatable, Sendable, Codable {
    var name: String
    var quantity: String
    var unit: String
    var amount: String

    init(name: String, quantity: String = "", unit: String = "", amount: String? = nil) {
        self.name = name
        self.quantity = quantity
        self.unit = unit
        self.amount = amount ?? RecipeIngredient.composeAmount(quantity, unit)
    }

    /// trim both; both empty -> ''; q empty -> u; u empty -> q; else "$q$u".
    static func composeAmount(_ quantity: String, _ unit: String) -> String {
        let q = quantity.trimmed
        let u = unit.trimmed
        if q.isEmpty && u.isEmpty { return "" }
        if q.isEmpty { return u }
        if u.isEmpty { return q }
        return "\(q)\(u)"
    }

    /// Legacy parse: trims, empty -> ('',''); uses `parseLeadingQuantity`:
    /// nil -> ('', wholeTrimmed); else (magnitude, remainder).
    private static func parseLegacyAmount(_ amount: String) -> (quantity: String, unit: String) {
        let trimmed = amount.trimmed
        if trimmed.isEmpty { return ("", "") }
        guard let parsed = QuantityText.parseLeadingQuantity(trimmed) else {
            return ("", trimmed)
        }
        return (parsed.magnitude, parsed.remainder)
    }

    /// True when the quantity is a plain numeric magnitude `scaledBy` can scale.
    var isScalable: Bool { Double(quantity.trimmed) != nil }

    /// Multiplies the numeric magnitude by `factor`, recomposing amount.
    /// `factor == 1` is a no-op preserving an explicit amount; a non-numeric
    /// quantity is returned unchanged.
    func scaledBy(_ factor: Double) -> RecipeIngredient {
        if factor == 1 { return self }
        guard let magnitude = Double(quantity.trimmed) else { return self }
        return RecipeIngredient(
            name: name,
            quantity: QuantityText.formatQuantity(magnitude * factor),
            unit: unit
        )
    }

    func copyWith(
        name: String? = nil,
        quantity: String? = nil,
        unit: String? = nil,
        amount: String? = nil
    ) -> RecipeIngredient {
        let preservedAmount = amount ?? (quantity == nil && unit == nil ? self.amount : nil)
        return RecipeIngredient(
            name: name ?? self.name,
            quantity: quantity ?? self.quantity,
            unit: unit ?? self.unit,
            amount: preservedAmount
        )
    }

    private enum CodingKeys: String, CodingKey { case name, quantity, unit, amount }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(name, forKey: .name)
        try c.encode(quantity, forKey: .quantity)
        try c.encode(unit, forKey: .unit)
        try c.encode(amount, forKey: .amount)
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let amount = c.decodeLenientIfPresent(String.self, forKey: .amount) ?? ""
        // "New shape" when either quantity or unit key is present.
        let hasNewShape = c.contains(.quantity) || c.contains(.unit)
        let name = c.decodeLenientIfPresent(String.self, forKey: .name) ?? ""
        if hasNewShape {
            self.init(
                name: name,
                quantity: c.decodeLenientIfPresent(String.self, forKey: .quantity) ?? "",
                unit: c.decodeLenientIfPresent(String.self, forKey: .unit) ?? "",
                amount: amount
            )
        } else {
            let parts = RecipeIngredient.parseLegacyAmount(amount)
            self.init(name: name, quantity: parts.quantity, unit: parts.unit, amount: amount)
        }
    }
}

/// De-duplicates by case-insensitive trimmed name, keeping the FIRST occurrence.
/// Must run at EVERY recipe entry point (matches `shoppingItemNameKey`).
func dedupeRecipeIngredients(_ ingredients: [RecipeIngredient]) -> [RecipeIngredient] {
    var seen = Set<String>()
    var result: [RecipeIngredient] = []
    for ingredient in ingredients {
        if seen.insert(ingredient.name.trimmed.lowercased()).inserted {
            result.append(ingredient)
        }
    }
    return result
}

/// Recipe entity. Identity (Hashable/Equatable) is by `id` ONLY.
struct Recipe: Hashable, Sendable, Codable {
    var id: String
    var name: String
    var category: String
    var difficulty: Int
    var cookingMinutes: Int
    var description: String
    var ingredients: [RecipeIngredient]
    var steps: [String]
    var tags: [String]
    var imageUrl: String?
    var remoteVersion: Int
    var clientUpdatedAt: Date?
    var deletedAt: Date?

    var syncMetadata: SyncMetadata {
        SyncMetadata(
            remoteVersion: remoteVersion,
            clientUpdatedAt: clientUpdatedAt,
            deletedAt: deletedAt
        )
    }

    /// `'难度未设置'` when difficulty <= 0, else `'难度 N/5'` with N clamped 1...5.
    var difficultyLabel: String {
        if difficulty <= 0 { return "难度未设置" }
        let level = min(max(difficulty, 1), 5)
        return "难度 \(level)/5"
    }

    init(
        id: String,
        name: String,
        category: String,
        difficulty: Int,
        cookingMinutes: Int,
        description: String,
        ingredients: [RecipeIngredient],
        steps: [String],
        tags: [String] = [],
        imageUrl: String? = nil,
        remoteVersion: Int = 0,
        clientUpdatedAt: Date? = nil,
        deletedAt: Date? = nil
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.difficulty = difficulty
        self.cookingMinutes = cookingMinutes
        self.description = description
        // Dedupe at the value-type entry point too (every entry point routes here).
        self.ingredients = dedupeRecipeIngredients(ingredients)
        self.steps = steps
        self.tags = tags
        self.imageUrl = imageUrl
        self.remoteVersion = remoteVersion
        self.clientUpdatedAt = clientUpdatedAt
        self.deletedAt = deletedAt
    }

    static func == (lhs: Recipe, rhs: Recipe) -> Bool { lhs.id == rhs.id }
    func hash(into hasher: inout Hasher) { hasher.combine(id) }

    private enum CodingKeys: String, CodingKey {
        case id, name, category, difficulty, cookingMinutes, description
        case ingredients, steps, tags, imageUrl
        case remoteVersion, clientUpdatedAt, deletedAt
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(name, forKey: .name)
        try c.encode(category, forKey: .category)
        try c.encode(difficulty, forKey: .difficulty)
        try c.encode(cookingMinutes, forKey: .cookingMinutes)
        try c.encode(description, forKey: .description)
        try c.encode(ingredients, forKey: .ingredients)
        try c.encode(steps, forKey: .steps)
        try c.encode(tags, forKey: .tags)
        try c.encodeAlways(imageUrl, forKey: .imageUrl)
        try c.encode(remoteVersion, forKey: .remoteVersion)
        try c.encodeISODateAlways(clientUpdatedAt, forKey: .clientUpdatedAt)
        try c.encodeISODateAlways(deletedAt, forKey: .deletedAt)
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let rawIngredients = c.decodeLenientIfPresent([RecipeIngredient].self, forKey: .ingredients) ?? []
        self.init(
            id: c.decodeLenientIfPresent(String.self, forKey: .id) ?? "",
            name: c.decodeLenientIfPresent(String.self, forKey: .name) ?? "",
            category: c.decodeLenientIfPresent(String.self, forKey: .category) ?? "",
            difficulty: c.decodeIntIfPresent(forKey: .difficulty) ?? 0,
            cookingMinutes: c.decodeIntIfPresent(forKey: .cookingMinutes) ?? 30,
            description: c.decodeLenientIfPresent(String.self, forKey: .description) ?? "",
            ingredients: rawIngredients,
            steps: c.decodeLenientIfPresent([String].self, forKey: .steps) ?? [],
            tags: c.decodeLenientIfPresent([String].self, forKey: .tags) ?? [],
            imageUrl: c.decodeLenientIfPresent(String.self, forKey: .imageUrl),
            remoteVersion: c.decodeIntIfPresent(forKey: .remoteVersion) ?? 0,
            clientUpdatedAt: c.decodeISODateIfPresent(forKey: .clientUpdatedAt),
            deletedAt: c.decodeISODateIfPresent(forKey: .deletedAt)
        )
    }

    func copyWith(
        id: String? = nil,
        name: String? = nil,
        category: String? = nil,
        difficulty: Int? = nil,
        cookingMinutes: Int? = nil,
        description: String? = nil,
        ingredients: [RecipeIngredient]? = nil,
        steps: [String]? = nil,
        tags: [String]? = nil,
        imageUrl: String? = nil,
        remoteVersion: Int? = nil,
        clientUpdatedAt: Date? = nil,
        deletedAt: Date? = nil,
        clearClientUpdatedAt: Bool = false,
        clearDeletedAt: Bool = false
    ) -> Recipe {
        Recipe(
            id: id ?? self.id,
            name: name ?? self.name,
            category: category ?? self.category,
            difficulty: difficulty ?? self.difficulty,
            cookingMinutes: cookingMinutes ?? self.cookingMinutes,
            description: description ?? self.description,
            ingredients: ingredients ?? self.ingredients,
            steps: steps ?? self.steps,
            tags: tags ?? self.tags,
            imageUrl: imageUrl ?? self.imageUrl,
            remoteVersion: remoteVersion ?? self.remoteVersion,
            clientUpdatedAt: clearClientUpdatedAt ? nil : (clientUpdatedAt ?? self.clientUpdatedAt),
            deletedAt: clearDeletedAt ? nil : (deletedAt ?? self.deletedAt)
        )
    }
}

extension String {
    /// Dart `.trim()` parity — strips leading/trailing whitespace & newlines.
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}
